hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/request/validation/ValidatorRegistry.java

```diff
@@ -0,0 +1,278 @@
+      registerValidator(requestType, allowedPhases, validator, reflections);
+    }
+  }
+
+
+
+  private void registerValidator(Class<RequestType> requestType,
+                                 Set<RequestProcessingPhase> allowedPhases,
+                                 Class<? extends Annotation> validatorToBeRegistered,
+                                 Reflections reflections) {
+    Collection<Method> methods =  reflections.getMethodsAnnotatedWith(validatorToBeRegistered);
+    Class<? extends Version> versionClass = (Class<? extends Version>)
+        this.getReturnTypeOfAnnotationMethod(validatorToBeRegistered, RegisterValidator.MAX_VERSION_METHOD_NAME);
+    List<Pair<? extends Annotation, Method>> sortedMethodsByMaxVersion = methods.stream()
+        .map(method -> Pair.of(method.getAnnotation(validatorToBeRegistered), method))
+        .sorted((validatorMethodPair1, validatorMethodPair2) ->
+            Integer.compare(
+                this.getMaxVersion(validatorMethodPair1.getKey()).getVersion(),
+                this.getMaxVersion(validatorMethodPair2.getKey()).getVersion()))
+        .collect(Collectors.toList());
+    for (Pair<? extends Annotation, Method> validatorMethodPair : sortedMethodsByMaxVersion) {
+      Annotation validator = validatorMethodPair.getKey();
+      Method method = validatorMethodPair.getValue();
+      Version maxVersion = this.getMaxVersion(validator);
+      RequestProcessingPhase phase = this.getRequestPhase(validator);
+      checkAllowedAnnotationValues(allowedPhases, phase, RegisterValidator.PROCESSING_PHASE_METHOD_NAME,
+          method.getName());
+      RequestType type = this.getRequestType(validator, requestType);
+      method.setAccessible(true);
+
+      EnumMap<RequestType, EnumMap<RequestProcessingPhase, IndexedItems<Method, Integer>>> requestMap =
+          this.indexedValidatorMap.get(versionClass);
+      EnumMap<RequestProcessingPhase, IndexedItems<Method, Integer>> phaseMap =
+          this.getAndInitialize(type, () -> new EnumMap<>(RequestProcessingPhase.class), requestMap);
+      this.getAndInitialize(phase, IndexedItems::new, phaseMap).add(method, maxVersion.getVersion());
+    }
+
+  }
+
+  private <K, V> V getAndInitialize(K key, Supplier<V> defaultSupplier, Map<K, V> from) {
+    return from.computeIfAbsent(key, k -> defaultSupplier.get());
+  }
+
+  static final class IndexedItems<T, IDX extends Comparable<IDX>> {
+    private final List<T> items;
+    private final TreeMap<IDX, Integer> indexMap;
+
+    private IndexedItems() {
+      this.items = new ArrayList<>();
+      this.indexMap = new TreeMap<>();
+    }
+
+    /**
+     * Add an item to the collection and update index if required. The order of items added should have their index
+     * sorted in increasing order.
+     * @param item
+     * @param idx
+     */
+    public void add(T item, IDX idx) {
+      indexMap.putIfAbsent(idx, items.size());
+      items.add(item);
+    }
+
+    /**
+     *
```

Comment:
Apidoc here would be nice, however I am fine if this method and the add method is not documented, however in this case please explain the IndexedItems class, and its role in the data structure within the registry somewhere.